File: python/Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def dfs(self, root, answer):
        if root:
            self.dfs(root.left, answer)
            answer.append(root.val)
            self.dfs(root.right, answer)

    def kthSmallest_rec(self, root, k):
        answer = []
        self.dfs(root, answer)
        return answer[k - 1]
    
    def kthSmallest(self, root, k):
        stack = []
        node = root
        while stack or node:
            if node:
                stack.append(node)
                node = node.left
            else:
                node = stack.pop()
                if k > 1:
                    k -= 1
                else:
                    return node.val
                
                node = node.right

File: cpp/Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> res;
    int count;

    int kthSmallest(TreeNode* root, int k) {
        res.clear();
        count = k;
        helper(root);
        return res[k - 1];
    }
    
    void helper(TreeNode* root) {
        if (root == NULL) {
            return;
        }
        if (res.size() < count) {
            helper(root->left);
            res.push_back(root->val);
            helper(root->right);
        }
    }
};
